core/src/providers/openai.rs

```diff
@@ -2008,72 +2012,98 @@ impl LLM for OpenAILLM {
+        let model_is_o1 = self.id.as_str().starts_with("o1-");
+
+        let (c, request_id) = if !model_is_o1 && is_streaming {
+            streamed_chat_completion(
+                self.chat_uri()?,
+                self.api_key.clone().unwrap(),
+                openai_org_id,
+                Some(self.id.clone()),
+                &openai_messages,
+                tools,
+                tool_choice,
+                temperature,
+                match top_p {
+                    Some(t) => t,
+                    None => 1.0,
+                },
+                n,
+                stop,
+                max_tokens,
+                match presence_penalty {
+                    Some(p) => p,
+                    None => 0.0,
+                },
+                match frequency_penalty {
+                    Some(f) => f,
+                    None => 0.0,
+                },
+                response_format,
+                openai_user,
+                event_sender.clone(),
+            )
+            .await?
+        } else {
+            chat_completion(
+                self.chat_uri()?,
+                self.api_key.clone().unwrap(),
+                openai_org_id,
+                Some(self.id.clone()),
+                &openai_messages,
+                tools,
+                tool_choice,
+                temperature,
+                match top_p {
+                    Some(t) => t,
+                    None => 1.0,
+                },
+                n,
+                stop,
+                max_tokens,
+                match presence_penalty {
+                    Some(p) => p,
+                    None => 0.0,
+                },
+                match frequency_penalty {
+                    Some(f) => f,
+                    None => 0.0,
+                },
+                response_format,
+                openai_user,
+            )
+            .await?
         };
 
+        // [o1-preview] Hack for OpenAI `o1-*` models to simulate streaming.
+        if model_is_o1 && is_streaming {
+            let sender = event_sender.as_ref().unwrap();
+            for choice in &c.choices {
+                if let Some(content) = &choice.message.content {
+                    // Split content into smaller chunks to simulate streaming.
+                    for chunk in content
+                        .chars()
+                        .collect::<Vec<char>>()
+                        .chunks(4)
+                        .map(|c| c.iter().collect::<String>())
+                    {
+                        let _ = sender.send(json!({
+                            "type": "tokens",
+                            "content": {
+                                "text": chunk,
+                            },
+                        }));
+                        // Add a small delay to simulate real-time streaming.
+                        tokio::time::sleep(std::time::Duration::from_millis(50)).await;
```

Comment:
as per IRL, a tiny bit faster might be nice

Reply:
Done!